library/core/src/ptr/non_null.rs

```diff
@@ -529,6 +529,11 @@ impl<T: ?Sized> NonNull<T> {
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
     #[stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_const_stable(feature = "non_null_convenience", since = "1.80.0")]
+    #[requires(
```

Comment:
I believe it's safe to call this function with `count == 0` even if `self.pointer` is dangling.

Reply:
> I believe it's safe to call this function with `count == 0` even if `self.pointer` is dangling.

Appreciate the suggestion @celinval! But there is some failure with the my dangling proof:

Proof harness:
```
#[kani::proof_for_contract(NonNull::byte_add)]
    pub fn non_null_byte_add_dangling_proof() {
        let ptr = NonNull::<i32>::dangling();
        let count: usize = kani::any();
        unsafe {
            let _ = ptr.byte_add(count);
        }
    }
```

Error:
```
VERIFICATION RESULT:
 ** 1 of 178 failed
Failed Checks: Kani does not support reasoning about pointer to unallocated memory
 File: "/Users/danieltu/Developer/verify-rust-std/library/core/src/lib.rs", line 423, in kani::mem::cbmc::same_allocation::<i32>

VERIFICATION:- FAILED
Verification Time: 0.26384938s
```

I am not sure whether it is because I didn't create a proper dangling scenario, could you provide some suggestion on this? Thanks!

Reply:
Did you extend the requires clause?

Reply:
> Did you extend the requires clause?

Hi @celinval, do you mind being more specific about it? Based on my understanding, we need to add a proof harness where `ptr` is dangling and `count` = 0. Any additional contract need to be added into the API?

Reply:
I think what @celinval meant is to change:
```rust
kani::mem::same_allocation(self.as_ptr(), self.as_ptr().wrapping_byte_offset(count))
```
to:
```rust
(count == 0 || kani::mem::same_allocation(self.as_ptr(), self.as_ptr().wrapping_byte_offset(count)))
```
so that `same_allocation` is not invoked if `count` is zero.